README.md
# Weathr — A Current Weather Application for Windows, Mac & Linux

This is an open-source current weather forecaster using OpenWeatherMaps and PyQt5 library.
It is written in Python 3. If you are installing from source, make sure to sign up for
[OpenWeatherMap](https://openweathermap.org/ "OpenWeatherMap") and request a key.

## Install from Source

Currently their is not actual file to download and install
currently, so you have to download from the source. To do so, use the 
command:

```
$ git clone https://github.com/henrybungee/weathr-forecast/
```

And you can compile from the source code with an IDE
of choice or run the MainWindow.py file. Or simply 
download the repository and run the files.

**I PROMISE THAT I WILL RELEASE AN ACTUAL FILE SOON**

But for now, if you want a copy, just run from the source
or use [PyInstaller](https://www.pyinstaller.org/ "PyInstaller's Homepage")
to compile the MainWindow.py file to a .exe, a .AppImage
or a .app. 

## Features

Weathr is a simple tool. But I can give you
the basic information you need to operate the 
software. For starters, the first time you boot
up, your city should be set to New York. 

![screeshot](https://github.com/henrybungee/weathr-forecast/blob/master/Screenshot%20from%202019-08-30%2011-12-12.png "Screenshot")

To change the city, click on the button that says
"Change City" (Simple enough, huh?). 
When you open the dialog, it's quite self-explanatory.
You have the option of selecting your own city or 
tracking location instead.

**NEW IN 2.6.3:**
UV tracking has been added. There are 5 levels of 
intensity. LOW, MODERATE, HIGH, VERY HIGH, and
EXTREME. Each of these has their own colors. 

   **Note:**
The location tracking is dependent on 
the wifi. If your router's location is 
not the same as where you are, you are 
out of luck.

Also, you can change conversions. By default,
it is the metric system conversions (Celsius,
KPH, etc.) but you can chage to the metric system
and even the ISU (International System of Units).

#### Shortcuts

Change City: Ctrl+Shift+C
Preferences: Ctrl+P

## Conclusion

Weathr is under the GPL v3.0 and all the rules are
to be followed. The breaking of those rules are 
against the GPL. Otherwise, have fun with this little
software and continue to make changes. Also, I think
I need help making the code more efficient.


CHANGELOG
## [2.6.3]

### Added
 — UV Index Display
 — UV Index Label changes color according to level.
 — Few Bug Fixes

src/NoInternet.py
from PyQt5 import QtGui, QtCore, QtWidgets
from PyQt5.QtGui import QPixmap, QIcon
from PyQt5.QtWidgets import QPushButton, QLabel, QMessageBox, QRadioButton, QLineEdit, QCheckBox

class NoInternetWindow(QtWidgets.QDialog):
    def __init__(self):
        QtWidgets.QWidget.__init__(self)
        self.setWindowTitle('Internet, Please?')
        self.top = 200
        self.left = 350
        self.width = 500
        self.height = 400
        #self.mainWindow = mainWindow
        self.cityName = ''

        self.InitUI()
        self.setGeometry(self.left, self.top, self.width, self.height)

    def InitUI(self):
        self.label = QLabel(self)
        self.label.setText('<h3>No Internet!</h3>')
        self.label.move(200, 20)

        self.noInternetImage = QPixmap('Images/nointernet.png')
        self.imageLabel = QLabel(self)
        self.imageLabel.setPixmap(self.noInternetImage)
        self.imageLabel.move(180, 80)
        self.explainationLabel = QLabel(self)
        self.explainationLabel.setWordWrap(True)
        self.explainationLabel.setText('Your Internet is down. Please try reconnecting to the internet.')
        self.explainationLabel.resize(300, 60)
        self.explainationLabel.move(110, 200)

        #create the close button
        self.closeBtn = QPushButton(self)
        self.closeBtn.setText('Close')
        self.closeBtn.move(15, 360)
        self.closeBtn.setStyleSheet('background-color: #2a42f7;')
        self.closeBtn.clicked.connect(self.close)

src/AboutWindow.py
from PyQt5 import QtGui, QtCore, QtWidgets, Qt
from PyQt5.QtWidgets import QLabel
#import MainWindow as main

class AboutWindow(QtWidgets.QDialog):
    def __init__(self):
        QtWidgets.QDialog.__init__(self)
        self.setWindowTitle('About Weathr')

        self.top = 220
        self.left = 430
        self.width = 500
        self.height = 350

        self.setGeometry(self.left, self.top, self.width, self.height)

        #define version etc.:
        self.VERSION = '2.6.3'
        self.VERSION_TYPE = 'Alpha'
        self.DESCRIPTION = 'Weathr is a current weather application written in Python using the Qt 5 library and OpenWeatherMap. With this application you have the power to see the current weather, as well as change your location and metric conversions.'

        self.labelTitle = QLabel(self)
        self.labelTitle.move(197, 20)
        self.labelTitle.setText("<h3>About Weathr</h3>")
        self.labelTitle.setStyleSheet('font-weight:bold;')

        self.versionInfo = QLabel(self)
        self.versionInfo.move(15, 75)
        self.versionInfo.setText('Version ' +self.VERSION)

        self.stable = QLabel(self)
        self.stable.setText(self.VERSION_TYPE)
        self.stable.move(15, 95)

        self.creatorName = QLabel(self)
        self.creatorName.setText('Created by Henry Lunger')
        self.creatorName.move(15, 115)

        self.emailMe = QLabel(self)
        self.emailMe.setWordWrap(True)
        self.emailMe.setText(self.DESCRIPTION)
        self.emailMe.move(15, 155)

        self.legalNote = QLabel(self)
        self.legalNote.setText('Powered by OpenWeatherMap\nCopyright© Henry Lunger 2019 CE')
        self.legalNote.move(15, 300)
        self.legalNote.setStyleSheet('color: grey;')

src/ConversionsWindow.py
from PyQt5 import QtGui, QtCore, QtWidgets
from PyQt5.QtWidgets import QVBoxLayout, QGroupBox, QPushButton, QLabel, QMessageBox, QRadioButton, QLineEdit
import MainWindow as main

class ConversionsWindow(QtWidgets.QDialog):
    def __init__(self, mainWindow):
        QtWidgets.QWidget.__init__(self)
        self.setWindowTitle('Preferences — Weathr')

        self.top = 200
        self.left = 350
        self.width = 600
        self.height = 430
        self.mainWindow = mainWindow

        self.setGeometry(self.left, self.top, self.width, self.height)

        self.labelTitle = QLabel(self)
        self.labelTitle.move(197, 20)
        self.labelTitle.setText("<h3>Weather Preferences</h3>")
        self.labelTitle.setStyleSheet('font-weight:bold;')

        #THE TEMPERATURE GROUP BOX

        #create the groupbox
        self.temperatureGroupbox = QGroupBox('Temperature Units', self)
        self.temperatureGroupbox.move(15, 50)

        #create the temperature radio buttons
        self.cRadioTemp = QRadioButton("Celsius (Metric System)", self)
        self.fRadioTemp = QRadioButton("Fahrenheit (Imperial System)", self)
        self.kRadioTemp = QRadioButton("Kelvin (Standard International Unit)", self)

        self.cRadioTemp.setChecked(True)
        if self.cRadioTemp.isChecked():
            self.unitChoice = 'C'

        #create the temp vbox
        self.vboxTemp = QVBoxLayout(self)
        self.vboxTemp.addWidget(self.cRadioTemp)
        self.vboxTemp.addWidget(self.fRadioTemp)
        self.vboxTemp.addWidget(self.kRadioTemp)
        self.vboxTemp.addStretch(1)

        #add items to the temp vbox
        self.temperatureGroupbox.setLayout(self.vboxTemp)

        #check the correct radio button
        if self.mainWindow.unitChoice == 'C':
            self.cRadioTemp.setChecked(True)
        elif self.mainWindow.unitChoice == 'F':
            self.fRadioTemp.setChecked(True)
        elif self.mainWindow.unitChoice == 'K':
            self.kRadioTemp.setChecked(True)

        #THE SPEED GROUP BOX

        #create the speed groupbox
        self.speedGroupbox = QGroupBox('Speed Units', self)
        self.speedGroupbox.move(15, 200)

        #create the radio buttons for groupbox
        self.kphRadio = QRadioButton('Kilometers Per Hour (Metric System)', self)
        self.mphRadio = QRadioButton('Miles Per Hour (Imperial System)', self)

        self.kphRadio.setChecked(True)

        #create the speed vbox
        self.speedVbox = QVBoxLayout(self)
        self.speedVbox.addWidget(self.kphRadio)
        self.speedVbox.addWidget(self.mphRadio)

        #add items to the groupbox
        self.speedGroupbox.setLayout(self.speedVbox)

        # check the correct speed radio button
        if self.mainWindow.speedUnitChoice == 'KPH':
            self.kphRadio.setChecked(True)
        elif self.mainWindow.speedUnitChoice == 'MPH':
            self.mphRadio.setChecked(True)

        self.saveBtn = QPushButton(self)
        self.saveBtn.setText('Save Preferences')
        self.saveBtn.move(15, 390)
        self.saveBtn.clicked.connect(self.checkRadioButtons)
        self.saveBtn.clicked.connect(self.reloadWeather)

        self.cancelBtn = QPushButton(self)
        self.cancelBtn.setText('Cancel')
        self.cancelBtn.move(485 ,390)
        self.cancelBtn.clicked.connect(self.reloadWeather)

    def reloadWeather(self):
        self.mainGetWeather = main.Window()
        self.mainGetWeather.getWeather()
        self.close()

    def checkRadioButtons(self):

        f = open('src/saves/speed.txt', 'r+')
        f.truncate(0)

        ftemp = open('src/saves/temp.txt', 'r+')
        ftemp.truncate(0)

        #save data
        if self.kphRadio.isChecked():
            with open("src/saves/speed.txt", "w") as f:
                f.write("KPH")
            self.mainWindow.speedUnitChoice = 'KPH'
        elif self.mphRadio.isChecked():
            with open("src/saves/speed.txt", "w") as f:
                f.write("MPH")
            self.mainWindow.speedUnitChoice = 'MPH'

        #save temperature unit data
        if self.cRadioTemp.isChecked():
            print("Celsius")
            with open("src/saves/temp.txt", "w") as f:
                f.write('C')
            self.mainWindow.unitChoice = 'C'
        elif self.fRadioTemp.isChecked():
            print("Fahrenheit")
            with open("src/saves/temp.txt", "w") as f:
                f.write('F')
            self.mainWindow.unitChoice = 'F'
        elif self.kRadioTemp.isChecked():
            print("Kelvin")
            with open("src/saves/temp.txt", "w") as f:
                f.write('K')
            self.mainWindow.unitChoice = 'K'

        #deals with speed XD
        if self.kphRadio.isChecked():
            print("KPH")
            self.mainWindow.speedUnitChoice = 'KPH'

        elif self.mphRadio.isChecked():
            print("MPH")
            self.mainWindow.speedUnitChoice = 'MPH'

src/ChangeCityWindow.py
from PyQt5 import QtGui, QtCore, QtWidgets
from PyQt5.QtGui import QIcon
from PyQt5.QtWidgets import QPushButton, QLabel, QMessageBox, QRadioButton, QLineEdit, QCheckBox

import MainWindow as main
import geocoder

class ChangeCityWindow(QtWidgets.QDialog):
    def __init__(self):
        QtWidgets.QWidget.__init__(self)
        self.setWindowTitle('Change The City — Weathr')
        self.top = 200
        self.left = 350
        self.width = 400
        self.height = 400
        self.cityName = ''

        self.InitUI()
        self.setWindowIcon(QIcon('Images/weather.ico'))
        self.setGeometry(self.left, self.top, self.width, self.height)


    def InitUI(self):

        #create a styled header
        self.header = QLabel(self)
        self.header.setText('<h3>Change the City</h3>')
        self.header.move(130, 10)

        #create the form and the explination label
        self.cityLabel = QLabel(self)
        self.cityLabel.setText('Change the City:')
        self.cityLabel.move(15, 30)

        #create input field
        self.cityInputField = QLineEdit(self)
        self.cityInputField.resize(180, 32)
        self.cityInputField.move(15, 55)
        self.cityInputField.setPlaceholderText('Enter a City...')

        #create header for settings
        self.headerSettings = QLabel(self)
        self.headerSettings.setText('<h4>Location Settings</h4>')
        self.headerSettings.move(15, 100)

        #create settings for the location settings
        self.locationLabel = QLabel(self)
        self.locationLabel.setText('NOTE: Location tracking is dependent on the internet and the IP. Note that the IP might not contain an accurate location. Also, if this is checked, whatever you put into the text field above will be ignored.')
        self.locationLabel.setWordWrap(True)
        self.locationLabel.move(15, 130)

        self.turnOnLocationTracking = QCheckBox(self)
        self.turnOnLocationTracking.setText('Track location instead')
        self.turnOnLocationTracking.move(15, 220)
        self.turnOnLocationTracking.stateChanged.connect(self.getLocation)
        self.turnOnLocationTracking.stateChanged.connect(self.cityInputField.setDisabled)

        self.coordinatesLabel = QLabel(self)
        self.coordinatesLabel.move(15, 245)
        self.coordinatesLabel.setText('<h4>Your Location will Be Displayed Here</h4>')
        self.coordinatesLabel.resize(600, 30)

        #create the save btn
        self.saveBtn = QPushButton(self)
        self.saveBtn.setText('Save Preferences')
        self.saveBtn.move(15, 360)
        self.saveBtn.clicked.connect(self.saveFunction)

        #create the cancel Btn
        self.cancelBtn = QPushButton(self)
        self.cancelBtn.setText('Cancel')
        self.cancelBtn.move(300, 360)
        self.cancelBtn.clicked.connect(self.cancel)


    def getLocation(self):

        self.g = geocoder.ip('me')
        print(self.g.latlng)
        self.coordinatesLabel.setText('<h4>Your Location Is In ' +str(self.g.state)+ ', ' +str(self.g.city)+ '</h4>')


    def saveFunction(self):

        #self.reloadWeather()
        #self.close()

        #check "weather" or not (hahaha get it? weather?) the location button is checked
        if self.turnOnLocationTracking.isChecked():

            #write to the text file
            with open('src/saves/startupcity.txt', 'w') as f:
                f.write(self.g.city)

            self.reloadWeather()
            self.close()
            
        #run if location tracking isn't checked.
        elif self.turnOnLocationTracking.isChecked() != True:

            if self.cityInputField.text() == '' or self.cityInputField.text() == ' ':
                self.dlg = QMessageBox.critical(self, 'Blank Form', 'Fill in the form. The form cannont be blank',
                                                QMessageBox.Ok)
                self.cancelBtn.setDisabled(True)
                self.mainGetWeather.close()
                self.show()

            else:
                #write to text file
                # write to the text file
                with open('src/saves/startupcity.txt', 'w') as f:
                    f.write(self.cityInputField.text())

                self.reloadWeather()
                self.close()
    def reloadWeather(self):

        self.mainGetWeather = main.Window()
        self.mainGetWeather.getWeather()
        self.close()

    def cancel(self):
        self.mainGetWeather = main.Window()
        self.close()
        self.mainGetWeather.show()